### DISCUSSION TITLE
用onlyoffice现在实现了文档ppt、excel、word在线浏览和编辑，求助怎么设置可以直接保存

### DISCUSSION THREAD

**Discussion by zhushen12580** [AlistGo/alist#3899] [Q&A] [answered]
对接的是自己搭建在服务器上的onlyoffice服务，目前文件能浏览和编辑，但编辑后不能直接保存，有些苦恼。

**Comment by chenjuefei** [comment, marked as answer]
编辑模式麻烦很多~在alist中，实现的意义不大，我情愿下载后编辑。不过后来发现有一个好处，就是能共同编辑。某些情况下，这个功能很好用！
编辑模式除了需要edit.html外，还需要配置一个回调处理程序 详细可参阅[https://api.onlyoffice.com/zh/editors/callback](https://api.onlyoffice.com/zh/editors/callback)。我自己用到是PHP。

callback.php
```
<?php

if (($body_stream = file_get_contents("php://input"))===FALSE){
    echo "Bad Request";
}

$data = json_decode($body_stream, TRUE);

if ($data["status"] == 2){
    $downloadUri = $data["url"];
        
    if (($new_data = file_get_contents($downloadUri))===FALSE){
        echo "Bad Response";
    } else {
        file_put_contents($path_for_save, $new_data, LOCK_EX);
    }
}
echo "{\"error\":0}";

?>
```
edit.html
```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>OnlyOffice editor</title>
</head>
 
<body>
    <div id="placeholder"></div>
    <script type="text/javascript" src="https://documentserver/web-apps/apps/api/documents/api.js"></script>
    <script>
        function getQueryParamValue(name) {
            const searchParams = new URLSearchParams(window.location.search);
            return searchParams.get(name);
        }
        const url = decodeURIComponent(getQueryParamValue("src"));
        const fileName = url.substring(url.lastIndexOf('/') + 1, url.lastIndexOf('?') != -1 ? url.lastIndexOf('?') : url.length);
        const fileExtension = fileName.split('.').pop();
        
        function generateDocumentKey(url) {
            const allowedChars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._=';
            const fileName = url.substring(url.lastIndexOf('/') + 1);
            const timestamp = new Date().toISOString();
            let key = fileName + timestamp;
            key = encodeURIComponent(key).substr(0, 20);
            key = key.split('').filter(char => allowedChars.includes(char)).join('');
            return key;
        }

        const docEditor = new DocsAPI.DocEditor("placeholder", {
            "document": {
                "fileType": fileExtension,
                "permissions": {
                    "edit": true,
                    "comment": true,
                    "download": true,
                    "print": true,
                    "fillForms": true,
                },
                "title": fileName,
                "url": url,
                "key": generateDocumentKey(url)
            },
            "editorConfig": {
                "lang": "zh-CN",
                "mode" : "edit" ,
                "callbackUrl": "https://example.com/url-to-callback.ashx",
            },
            "height": "1080px",
            "type": "desktop",
        });
    </script>
</body>
</html>
```
`https://example.com/url-to-callback.ashx 和 https://documentserver/web-apps/apps/api/documents/api.js`需要修改成自己的

https://user-images.githubusercontent.com/35056864/227172150-2535ca0d-dd23-4b17-829d-eebfabf12e35.mp4

**Reply by zhushen12580** [reply]
膜拜 大佬

**Reply by zhushen12580** [reply]
> 编辑模式麻烦很多~在alist中，实现的意义不大，我情愿下载后编辑。不过后来发现有一个好处，就是能共同编辑。某些情况下，这个功能很好用！ 编辑模式除了需要edit.html外，还需要配置一个回调处理程序 详细可参阅https://api.onlyoffice.com/zh/editors/callback。我自己用到是PHP。
> 
> callback.php
> 
> ```
> <?php
> 
> if (($body_stream = file_get_contents("php://input"))===FALSE){
>     echo "Bad Request";
> }
> 
> $data = json_decode($body_stream, TRUE);
> 
> if ($data["status"] == 2){
>     $downloadUri = $data["url"];
>         
>     if (($new_data = file_get_contents($downloadUri))===FALSE){
>         echo "Bad Response";
>     } else {
>         file_put_contents($path_for_save, $new_data, LOCK_EX);
>     }
> }
> echo "{\"error\":0}";
> 
> ?>
> ```
> 
> edit.html
> 
> ```
> <!DOCTYPE html>
> <html lang="en">
> <head>
>     <meta charset="UTF-8">
>     <title>OnlyOffice editor</title>
> </head>
>  
> <body>
>     <div id="placeholder"></div>
>     <script type="text/javascript" src="https://documentserver/web-apps/apps/api/documents/api.js"></script>
>     <script>
>         function getQueryParamValue(name) {
>             const searchParams = new URLSearchParams(window.location.search);
>             return searchParams.get(name);
>         }
>         const url = decodeURIComponent(getQueryParamValue("src"));
>         const fileName = url.substring(url.lastIndexOf('/') + 1, url.lastIndexOf('?') != -1 ? url.lastIndexOf('?') : url.length);
>         const fileExtension = fileName.split('.').pop();
>         
>         function generateDocumentKey(url) {
>             const allowedChars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._=';
>             const fileName = url.substring(url.lastIndexOf('/') + 1);
>             const timestamp = new Date().toISOString();
>             let key = fileName + timestamp;
>             key = encodeURIComponent(key).substr(0, 20);
>             key = key.split('').filter(char => allowedChars.includes(char)).join('');
>             return key;
>         }
> 
>         const docEditor = new DocsAPI.DocEditor("placeholder", {
>             "document": {
>                 "fileType": fileExtension,
>                 "permissions": {
>                     "edit": true,
>                     "comment": true,
>                     "download": true,
>                     "print": true,
>                     "fillForms": true,
>                 },
>                 "title": fileName,
>                 "url": url,
>                 "key": generateDocumentKey(url)
>             },
>             "editorConfig": {
>                 "lang": "zh-CN",
>                 "mode" : "edit" ,
>                 "callbackUrl": "https://example.com/url-to-callback.ashx",
>             },
>             "height": "1080px",
>             "type": "desktop",
>         });
>     </script>
> </body>
> </html>
> ```
> 
> `https://example.com/url-to-callback.ashx 和 https://documentserver/web-apps/apps/api/documents/api.js`需要修改成自己的
> 
>  bandicam.2023-03-23.18-11-14-465.mp4

大佬 这个文件url-to-callback.ashx可以发一下吗 /(★ ω ★)

**Reply by homuraLan** [reply]
> 编辑模式麻烦很多~在alist中，实现的意义不大，我情愿下载后编辑。不过后来发现有一个好处，就是能共同编辑。某些情况下，这个功能很好用！ 编辑模式除了需要edit.html外，还需要配置一个回调处理程序 详细可参阅https://api.onlyoffice.com/zh/editors/callback。我自己用到是PHP。
> 
> callback.php
> 
> ```
> <?php
> 
> if (($body_stream = file_get_contents("php://input"))===FALSE){
>     echo "Bad Request";
> }
> 
> $data = json_decode($body_stream, TRUE);
> 
> if ($data["status"] == 2){
>     $downloadUri = $data["url"];
>         
>     if (($new_data = file_get_contents($downloadUri))===FALSE){
>         echo "Bad Response";
>     } else {
>         file_put_contents($path_for_save, $new_data, LOCK_EX);
>     }
> }
> echo "{\"error\":0}";
> 
> ?>
> ```
> 
> edit.html
> 
> ```
> <!DOCTYPE html>
> <html lang="en">
> <head>
>     <meta charset="UTF-8">
>     <title>OnlyOffice editor</title>
> </head>
>  
> <body>
>     <div id="placeholder"></div>
>     <script type="text/javascript" src="https://documentserver/web-apps/apps/api/documents/api.js"></script>
>     <script>
>         function getQueryParamValue(name) {
>             const searchParams = new URLSearchParams(window.location.search);
>             return searchParams.get(name);
>         }
>         const url = decodeURIComponent(getQueryParamValue("src"));
>         const fileName = url.substring(url.lastIndexOf('/') + 1, url.lastIndexOf('?') != -1 ? url.lastIndexOf('?') : url.length);
>         const fileExtension = fileName.split('.').pop();
>         
>         function generateDocumentKey(url) {
>             const allowedChars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._=';
>             const fileName = url.substring(url.lastIndexOf('/') + 1);
>             const timestamp = new Date().toISOString();
>             let key = fileName + timestamp;
>             key = encodeURIComponent(key).substr(0, 20);
>             key = key.split('').filter(char => allowedChars.includes(char)).join('');
>             return key;
>         }
> 
>         const docEditor = new DocsAPI.DocEditor("placeholder", {
>             "document": {
>                 "fileType": fileExtension,
>                 "permissions": {
>                     "edit": true,
>                     "comment": true,
>                     "download": true,
>                     "print": true,
>                     "fillForms": true,
>                 },
>                 "title": fileName,
>                 "url": url,
>                 "key": generateDocumentKey(url)
>             },
>             "editorConfig": {
>                 "lang": "zh-CN",
>                 "mode" : "edit" ,
>                 "callbackUrl": "https://example.com/url-to-callback.ashx",
>             },
>             "height": "1080px",
>             "type": "desktop",
>         });
>     </script>
> </body>
> </html>
> ```
> 
> `https://example.com/url-to-callback.ashx 和 https://documentserver/web-apps/apps/api/documents/api.js`需要修改成自己的
> 
>  bandicam.2023-03-23.18-11-14-465.mp4

我成功了给大佬点个赞 太厉害了

**Reply by zhushen12580** [reply]
能分享下经验吗

**Reply by homuraLan** [reply]
> 能分享下经验吗

https://github.com/Sincejunly/officeaanglist.git

**Comment by Alano-i** [comment]
@chenjuefei 大佬，请问能分享下url-to-callback.ashx文件吗，我用你上面那个callback.php提示无法保存，报下面这个错，感谢！

<img width="1618" alt="image" src="https://user-images.githubusercontent.com/68833595/233886080-3e5968c9-0ecc-423b-a806-7a2bf861f0a9.png">

**Reply by homuraLan** [reply]
> @chenjuefei 大佬，请问能分享下url-to-callback.ashx文件吗，我用你上面那个callback.php提示无法保存，报下面这个错，感谢！
> 
> <img alt="image" width="1618" src="https://user-images.githubusercontent.com/68833595/233886080-3e5968c9-0ecc-423b-a806-7a2bf861f0a9.png">

https://github.com/Sincejunly/officeaanglist.git

**Comment by toyqiu** [comment]
求教各位大佬：callback.php 文件应该放在哪个路径下

**Reply by homuraLan** [reply]
> 求教各位大佬：callback.php 文件应该放在哪个路径下

https://github.com/Sincejunly/officeaanglist.git

**Comment by theBig-brother** [comment]
你怎么做到实现了文档ppt、excel、word在线浏览和编辑的，能教一下吗，比如后端代码

**Comment by Li-Aaron** [comment]
根据大佬提供的思路
写了个比较粗鄙的callback.php
```php
<?php
function curl_post($url, $header, $data){
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_URL, $url);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, FALSE);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, FALSE);
    curl_setopt($curl, CURLOPT_POST, 1);
    curl_setopt($curl, CURLOPT_POSTFIELDS, $data);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $header);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
    $output = curl_exec($curl);
    curl_close($curl);
    return $output;
}

function curl_put($url, $header, $data){
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_URL, $url);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, FALSE);
    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, FALSE);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $header);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "PUT");
    curl_setopt($curl, CURLOPT_ENCODING, '');
    curl_setopt($curl, CURLOPT_MAXREDIRS, 10);
    curl_setopt($curl, CURLOPT_TIMEOUT, 0);
    curl_setopt($curl, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
    curl_setopt($curl, CURLOPT_POSTFIELDS, $data);

    $output = curl_exec($curl);
    curl_close($curl);
    return $output;
}

// 获取原始文件名
$input_file_url = $_GET['file_url'];
if (($body_stream = file_get_contents("php://input"))===FALSE){
    echo "{\"error\":1}";
}
$data = json_decode($body_stream, TRUE);

if ($data["status"] == 2){
    // 获取待传输文件，content保存在new_data
    $downloadUri = $data["url"];
    $stream_opts = [
        "ssl" => [
            "verify_peer"=>false,
            "verify_peer_name"=>false,
        ]
    ];

    if (($new_data = file_get_contents($downloadUri, false, stream_context_create($stream_opts)))===FALSE){
        echo "{\"error\":2}";
    } else {
        // POST token 获取
        $post_url = "http://alist/api/auth/login";
        $post_header = [
            "Content-type:application/json;charset='utf-8'",
            "Accept:application/json"
        ];
        $post_data = [
            'Username' => 'username',
            'Password' => 'password'
        ];
        $post_data = json_encode($post_data);
        $post_response = curl_post($post_url, $post_header, $post_data);
        $post_response_data = json_decode($post_response, true);
        $token = $post_response_data["data"]["token"];

        // 执行put操作 直接上传文件
        // 这里写的比较粗鄙
        $put_url = "http://alist/api/fs/put";
        $sublen = strlen("http://alist/p") - strlen($input_file_url);
        $filename = urlencode(substr($input_file_url, $sublen));
        $put_header = [
            sprintf("Authorization:%s", $token),
            sprintf("File-Path:%s", $filename),
            "Content-Type:application/octet-stream"
        ];
        $put_response = curl_put($put_url, $put_header, $new_data);
    }
}
echo "{\"error\":0}";

?>
```

大概流程就是通过php的get获取原始url，根据原始url解析文件名，post获取token，put上传文档。
内网可以凑合用用。

edit.html要相应改一下
```html
            "editorConfig": {
                "lang": "zh-CN",
                "mode" : "edit" ,
                "callbackUrl": "http://phphost/callback.php?file_url="+url,
            },
```

http://alist
http://phphost
以及用户名密码需要改一下。

php我这边用的是nginx+php-fpm.

**Reply by kun4399** [reply]
谢谢，成功了

**Reply by anwen-anyi** [reply]
好像如果开启了全部签名  保存后会有一点儿问题. :joy_cat: 
![image](https://github.com/alist-org/alist/assets/56105412/f6367ac3-06b6-42f1-89f2-bd8892084339)

**Reply by imysl0** [reply]
能编辑，也显示保存成功，就是alist那边没有拿到保存后的文件，put的问题？也没有看到哪里报错

**Reply by anwen-anyi** [reply]
> 能编辑，也显示保存成功，就是alist那边没有拿到保存后的文件，put的问题？也没有看到哪里报错

估计你和我上面那个问题一样 开启了全部签名 修改的文件携带了   `office.doc?sign=xxxxx` 参数

**Reply by imysl0** [reply]
我看了，没开启签名，就奇怪~

**Reply by imysl0** [reply]
最后用户名和密码那里设置为管理员用户名和密码后可以了，原因未知，难道是因为管理员和普通用户开始路径不同？我设置管理员能看到所有路径，普通用户只能看到某个子文件夹。

**Reply by imysl0** [reply]
所以内网用用这句话不是白说的，哈哈哈

**Reply by Li-Aaron** [reply]
确实我就一个管理员用户自己在用，没考虑那么多情况。（另外我也比较菜）
